Fix for pytest 4 and 5, support pythons 2.7-3.8, update CIs

pv opened the pull request:
Do some reorganization, update pytest-leaks for current pytest, and add support for different Python versions.

The largest reorganization here is to move the refcheck code to separate file(s), copied from cpython upstream with minimal patching (and changed lines marked), so that it's easier to keep it in sync with Python and support different Python versions. Since the refcheck code contains parts that patch cpython internals, it may need updating later on (as already happened with the `collections.abc` registry) which is made easier by keeping the patching minimal.

The pytest4/5 fix is simple (as mentioned by @seberg).

I also updated the CIs to use Python debug builds, so that they can actually check this (getting this to work was the part that took longest...)

I could maybe separate this in different PR, but not sure if it makes sense in project this small.

Closes: #4
Closes: #5
Closes: #6  (adds warning in documentation) 
Closes: #8
Closes: #10

pv committed:
Fix for pytest 4 and 5

pv committed:
Rename example file

pv committed:
Drop (incomplete) support for Python 2

pv committed:
Make pytest_runtest_protocol to report errors immediately

Log errors immediately, without proceeding to run standard pytest code.

pv committed:
Raise error if not running on Python debug build

pv committed:
Add refleak.py from Python 3.8

pv committed:
Edit refleak_38.py for pytest-leaks

Only changes the approach to reporting the results.

pv committed:
Reorganize as a Python package

pv committed:
Use the patched refleak_38.py

pv committed:
TST: fix tests for new pytest

pv committed:
Add abridged Lib/test/regrtest.py from py2.7

pv committed:
Modify refleak_27 for use here

pv committed:
STY: fix flake8 issues

pv committed:
DOC: update readme

pv committed:
Add abridged Lib/test/regrtest.py from py3.5

pv committed:
Modify refleak_35.py for use here

pv committed:
Fix operation on python 3.6

pv committed:
CI: update travis-CI config

pv committed:
CI: update appveyor

pv committed:
Better checking of ini/-R parameters

pv committed:
TST: tox.ini: update basepython name to match debian

pv committed:
Bump setup.py version

seberg commented:
Wow, a big PR, but I very much appreciate pytest-leaks moving forward. My personal take would be to just keep it moving, considering how broken it is for pytest 4 and 5, that is a much larger issue than possibly breaking someone (who can just pin the old version).

pv committed:
Fix handling of Ctrl-C

pv committed:
Work around issue with pytest-sugar

pv committed:
Prettify output formatting

pv committed:
CI: guard against partial appveyor cache restore

pv committed:
Fix leak checking for doctests

Work around an issue in missing globals when running the same doctest
multiple times, and ensure errors get formatted properly.

The pytest.doctest machinery itself appears to leak when run
with `testdir` fixture in the tests under some cases. We test
only leak detection, but it works outside pytest-leaks tests.

pv committed:
DOC: add notice about running a test multiple times

seberg commented:
Just to give a thumbs up, I am running the numpy test suit with this, and it seems to run nicely (it feels slower than previously, probably because the ABC stuff takes some cpu cycles, and I had that commented out brutally, but that seems OK).

abalkin commented:
@pv, @seberg  - I could use some help maintaining this package.  Would you like to join as collaborators?

seberg commented:
Cannot say I know much about it as of now, but I can try to help out a bit, it would be nice to just make sure the project does not die completely, it is very nice.

seberg commented on pytest_leaks/refleak_35.py:
```diff
@@ -0,0 +1,281 @@
+    # Collect cyclic trash and read memory statistics immediately after.
+    func1 = sys.getallocatedblocks
+    func2 = sys.gettotalrefcount
+    gc.collect()
```
I am curious if we should do `for i in range(20): if gc.collect() == 0: break` (possibly with 1-2 extra runs), or leave deep cyclic cleanup to the test writer (I do not care much either way, I had to do such a thing in numpy in one spot; And I suppose suppose it might slow down things a bit, since it will probably run at least twice usually, instead of once.)

pv replied:
The cpython>=3.7 refleak code indeed calls gc multiple (3) times.

Here, I didn't really change the cpython code at all, it's direct copypaste (modulo a few explicitly marked changes). I guess we can rationalize this when we know better, or combine the codebases for the different Python versions.

seberg replied:
Ah ok, I did not realize. Breaking early if 0 is reached may be a nice thing. I think Matti mentioned that at least in pypy, you need 2-3 extra calls (OTOH, I do not think pypy is of primary concern here).

pv commented:
@abalkin: yes, that'd be ok with me, I'd like to help this project stay alive.

seberg commented on pytest_leaks/refleak_38.py:
```diff
@@ -0,0 +1,293 @@
+    # Clear the warnings registry, so they can be displayed again
+    for mod in sys.modules.values():
+        if hasattr(mod, '__warningregistry__'):
+            del mod.__warningregistry__
```
Do we have to do this on newer python versions? On <3.5 it should be necessary, but for 3.7 and 3.8 I would hope not?

abalkin commented on pytest_leaks/__init__.py:
```diff
@@ -0,0 +1 @@
+__version__ = "0.3.dev"
```
I would bump this directly to 0.3.0 and upload to PyPI shortly after merge.

pv committed:
Bump to 0.3.0

abalkin commented on setup.py:
```diff
@@ -13,7 +13,7 @@ def read(fname):
 setup(
     name='pytest-leaks',
-    version='0.2.2',
+    version='0.3.dev',
```
Pauli - please feel free to add yourself as co-author.  Your contribution deserves recognition!

abalkin commented:
@pv - your work looks good to me.  Feel free to merge when you are ready.  Please let me know if the privileges that I gave you are not enough.

jayvdb commented on appveyor.yml:
```diff
@@ -1,33 +1,31 @@
+      }
+  # Install packages
+  - SET PYTHON=%LOCALAPPDATA%\cpython\PCbuild\win32\python_d.exe
+  - "%PYTHON% -mensurepip --user"
```
A space after `-m` aids reading these lines.

jayvdb commented on LICENSE:
```diff
@@ -20,3 +20,54 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+
+
+Furthermore, the files pytest_leaks/refleak_*.py and
+pytest_leaks/support.py are covered by the PSF license:
```
The short name for this license is `Python-2.0` https://spdx.org/licenses/Python-2.0.html

Sticking it in here will probably make GitHub's license detection fail.  I suggest having two files, LICENSE and LICENSE-PSF.txt

And then glue them together by mentioning them both in the setup.py : `license='MIT and Python-2.0'` , and possibly mention both in the README too, and also ensure the next release includes both license files in the sdist.

pv replied:
Thanks, going to postpone license splitting in separate PR

jayvdb replied:
As predicted, GitHub is now showing "View license" because it cant detect it.  Not terrible, but I think it is worth the effort to convince/trick GitHub that it is MIT license, because the addition of PSF license has zero impact on 99% of users because they are already already running CPython.  And the other 1% are distro packagers, and they should be checking the licenses in the sdist (and completely ignoring whatever GitHub says the license might be).

seberg replied:
This may be a silly question, but does the PSF come from the template? Because at this stage and number of authors, it would also not be an issue to simply relicense the code. Although, it is probably not really necessary.

pv replied:
PSF comes because refleak_*.py are copypasted code from cpython (and there was copypasted code also before this PR).
Since we're not PSF ourselves, the license probably won't make sense for code we write, so it's going to be dual license.

I'm actually not so convinced we should start doing things just to get the github badge right.

pv committed:
Update setup.py authors, license, requires, classifiers
